Scheme location simulation

freak4pc opened the pull request:
### Short description 📝

This PR adds a new `simulatedLocation` property on RunAction's options. 

This allows setting either one of Xcode's built-in simulated locations for a scheme, or provide your own GPX file (as long as it's in the project's resources).

For example, this is done using

```swift
runAction: RunAction(config: .debug,
                     executable: "QA",
                     arguments: Arguments(
                       environment: [:],
                       launchArguments: []
                     ),
                     options: .options(
                       storeKitConfigurationPath: nil,
                       simulatedLocation: .custom(gpxFile: "Gett/Resources/Habarzel 21.gpx")
                     )
)
```

<img width="940" alt="image" src="https://user-images.githubusercontent.com/605076/110255194-20177a00-7f9b-11eb-90c9-02d2dd39a9bb.png">


I've tested this locally and added a quick unit test for the "inner" part, but not entirely sure how to test it from the outer interface (ProjectDescription side) - would love some guidance on that.

I'd also appreciate some guidance on where should I update the documentation - is it just the MDX file or anything beyond that? 

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

kwridan commented on Sources/ProjectDescription/RunActionOptions.swift:
```diff
@@ -27,8 +35,113 @@ public struct RunActionOptions: Equatable, Codable {
+            self = .newYork
+        case "Rio De Janeiro, Brazil":
+            self = .rioDeJaneiro
+        case _ where value.contains("/"):
+            self = .custom(gpxFile: Path(value))
```
When using relative paths there may not always be a `/` in the final value:

e.g.

```swift
let userScheme = Scheme(name: "App-Local",
                        shared: false,
                        buildAction: BuildAction(targets: ["App"], preActions: [debugAction]),
                        testAction: TestAction(targets: ["AppTests"]),
                        runAction: RunAction(
                            executable: "App",
                            options: .options(
                                simulatedLocation: .custom(gpxFile: "MyGpxFile.gpx")
                            )
                        ))
```

This will yield an `CodingError.unknownLocation` coding error, encoding the type explicitly may help (a bit of hassle, but can remove ambiguity).

A unit test for the custom codable may be worthwhile, `RunActionOptions` doesn't have one yet, but you could copy any of the existing `ProjectDescriptionTests` as a starter for it.

An alternative to ease the codable dance is to cheat slightly and use a struct with optionals however expose the desired cases as memberwise static initialisers e.g.

```swift
public struct SimulatedLocation: Codable {
    public let identifier: String?
    public let gpxFile: Path?

    public static var london: SimulatedLocation {
        SimulatedLocation(identifier: "London, England", gpxFile: nil)
    }

    // ...

    public static func custom(gpxFile: Path) -> SimulatedLocation {
        SimulatedLocation(identifier: nil, gpxFile: gpxFile)
    }
}
```

The only so-so compromise is we have to assume we'll never have a case where both variables are defined or empty as we're not exposing the direct initializer (enums are supposed to be the solution to this, but as soon as you add associated types, you need to pay the codable tax!)

kwridan commented on Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift:
```diff
@@ -386,15 +387,35 @@ final class SchemeDescriptorsGenerator: SchemeDescriptorsGenerating {
+        {
+            var identifier = locationScenario.identifier
+
+            /// GPX file path is the relative path between the storekit file, and the xcode project
```
the comment may need an update (if its needed at all) 😅

kwridan reviewed (commented):
Nice work @freak4pc, you've found your way around the codebase already 👍 

It could be worth tweaking the fixture `ios_app_with_custom_scheme` to include an example of this new feature which can act as a show case for future users (we don't necessarily need to add an acceptance test for it). 

Ah and the docs in https://github.com/tuist/tuist/blob/main/website/markdown/docs/usage/project-description.mdx#run-action-options will need an update to include a new entry for this.

Thanks for adding this!

freak4pc commented:
Thanks @kwridan ! 

1. I've updated the docs
2. I've simplified the types similarly to your idea to get the free `Codable` conformance. I decided to go for `hasSuffix(".gpx")` for the path detection which seems reasonable.
3. It seems I can't really get any tests going, since it would require adding a new step definition. I started doing this but, [Xcodeproj.launch_action itself doesn't extract `LocationScenarioReference`](https://github.com/CocoaPods/Xcodeproj/blob/master/lib/xcodeproj/scheme/launch_action.rb#L39-L48) so I have no way to validate that (without a secondary pull request to XcodeProj, possibly).

Let me know if you have any additional thoughts on this. Thank you!

freak4pc commented:
@kwridan I've added the option to `ios_app_with_custom_scheme` as an example, hope that works ::)

freak4pc commented:
I have this PR up which will allow testing this feature more thoroughly at a later point:
https://github.com/CocoaPods/Xcodeproj/pull/813

fortmarek reviewed (approved):
Looks great, thanks for doing this @freak4pc! Let's add a more thorough integration test once the PR in CocoaPods is merged.

I'll let @kwridan make another round of review for this 🙂

kwridan reviewed (commented):
Thanks for the updates @freak4pc

> I've simplified the types similarly to your idea to get the free Codable conformance. I decided to go for hasSuffix(".gpx") for the path detection which seems reasonable.

Indeed this would have sufficed if `Path` is purely a string, but alas it's a bit more sophisticated, luckily it's Codable so should be easy to bridge it across without losing some of its contextual information.

> It seems I can't really get any tests going, since it would require adding a new step definition. I started doing this but, Xcodeproj.launch_action itself doesn't extract LocationScenarioReference so I have no way to validate that (without a secondary pull request to XcodeProj, possibly).

> I have this PR up which will allow testing this feature more thoroughly at a later point:
CocoaPods/Xcodeproj#813

Ah nice I'm sure it will come in handy in the future! apologies didn't mean we needed acceptance tests per se, having a fixture exercise the code path will implicitly get it tested even if there isn't any verification taking place on the desired side effect. For example, it will ensure that the generation continues to pass with / without it, will allow us to manually test it, lastly acts as a nice showcase should someone wonder how the feature is used in a real example :) - Thanks for adding the fixture 👍 

ps. to get the fixture running locally I typically do

```sh
swift build
swift run tuist generate --path fixtures/ios_app_with_custom_scheme
```

And then inspect the generated workspace.

---

While testing I spotted a few more bits:

### The gpx files need to be part of the project/workspace

I noticed the `.gpx` file needs to be included in the project/workspace to take effect. We can copy the approach taken by the store kit files and automatically add gpx files in [`ProjectFileElements`](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Generator/ProjectFileElements.swift#L112)


### The path to gpx files

 it seems the the scheme identifier for `.gpx` file is a bit more nuanced that first appears :(

The generated `App-Release.xcscheme` has:

```xml
     <LocationScenarioReference
         identifier = "../Resources/Grand Canyon.gpx"
         referenceType = "0">
      </LocationScenarioReference>
```

This doesn't appear to set the simulated location however in the Xcode UI 

![Screenshot 2021-03-09 at 07 45 34](https://user-images.githubusercontent.com/11914919/110435870-74118400-80ab-11eb-9ce2-986e45b03356.png)

Manually selecting in the drop down yields a slightly odd path in the scheme:

```xml
      <LocationScenarioReference
         identifier = "../App/Resources/Grand Canyon.gpx"
         referenceType = "0">
      </LocationScenarioReference>
```

It appears the path in the scheme is relative to the xcworkspace and not the Xcode project ... even though it's a project level scheme🤔

This seems to get it working:

```swift
if case let .gpxFile(gpxPath) = locationScenario {
                let fileRelativePath = gpxPath.relative(to: graphTraverser.workspace.xcWorkspacePath)
                identifier = fileRelativePath.pathString
            }
```

Some further testing shows it's a bit more annoying than that, it's dependent on how you view the project, within an xcworkspace vs a standalone xcodeproj. 

For the standalone xcodeproj case, when viewing it in Xcode, Xcode automatically generates an xcworkspace that resides within the `.xcodeporj` (`App.xcodeproj/project.xcworkspace`) and in that case it needs to be relative to that nested workspace to work:

```xml
      <LocationScenarioReference
         identifier = "../../Resources/Grand Canyon.gpx"
         referenceType = "0">
      </LocationScenarioReference>
```

This makes it tricky to deterministically generate the scheme 😞

I guess we can try and cater for the common case of the viewing the projects within the explicit workspace _(leveraging `graphTraverser.workspace.xcWorkspacePath`)_ - I'm surprised we didn't run into this when the store kit files support was introduced as I suspect it has the same path fun.

kwridan commented:
It's worth verifying some of my findings, I hope it's just a quirk of the specific Xcode version _(e.g. I was testing with 12.1)_

freak4pc commented on Sources/ProjectDescription/RunActionOptions.swift:
```diff
@@ -26,9 +34,79 @@ public struct RunActionOptions: Equatable, Codable {
+        }
+
+        public static func custom(gpxFile: Path) -> SimulatedLocation {
+            .init(identifier: gpxFile.pathString)
```
Makes sense! Thanks :)

freak4pc replied:
It does mean I have to use the two properties solution, like you suggested earlier, If I understand correctly? (One for path, one for identifier)

kwridan replied:
Ya that should do the trick

freak4pc commented:
Thanks for the great and details feedback @kwridan - I'll confirm that's indeed the case with our own project but if yes, I think supporting just the workspace case makes 100% sense. 

Also, I wanted to automatically copy the GPX into the project resources but wasn't sure how to, thanks for pointing me there.

I'll update this and ping you again soon ;)

freak4pc commented:
I think this is good to go @kwridan ! Thank you :)

kwridan reviewed (approved):
Awesome, thanks for the updates @freak4pc! 

The fixture now works for both cases, the gpx file and the predefined location 🎉
